src/mission_companion.cpp

```diff
@@ -1173,6 +1339,350 @@ bool talk_function::labor_return( npc &p )
     return true;
 }
 
+void talk_function::om_camp_upgrade( npc &comp ){
+    editmap edit;
+    const point omt_pos = ms_to_omt_copy( g->m.getabs( comp.posx(), comp.posy() ) );
+    oter_id &omt_ref = overmap_buffer.ter( omt_pos.x, omt_pos.y, comp.posz() );
+    std::string om_old = omt_ref.id().c_str();
+    if (om_old == "field"){
+        edit.set_mapgen( "faction_base_camp_1", comp.pos());
+    } else {
+        edit.set_mapgen( om_next_upgrade(om_old), comp.pos());
+    }
+
+    tripoint np = tripoint( 5 * SEEX + 5 % SEEX, 4 * SEEY + 5 % SEEY, comp.posz() );
+    np.x = g->u.posx()+1;
+    np.y = g->u.posy();
+
+    tripoint p( 0, 0, comp.posz() );
+    int &x = p.x;
+    int &y = p.y;
+    for( x = 0; x < SEEX * g->m.getmapsize(); x++ ) {
+        for( y = 0; y < SEEY * g->m.getmapsize(); y++ ) {
+            if(g->m.ter(x, y) == ter_id( "t_floor_green" )) {
+                np.x = x;
+                np.y = y;
+            }
+        }
+    }
+    comp.setpos(np);
+    comp.set_destination();
+    g->load_npcs();
+    popup(_("Upgrade completed."));
+}
+
+std::string talk_function::om_upgrade_description( npc &p, std::string bldg ){
+    const recipe *making = &recipe_id( bldg ).obj();
+    if( making == nullptr ) {
```

Comment:
This will *never* be `true`. `making` has been assigned the address of something (the result of applying the "address of" operator `&`). The address of something can never be null.

Note that `recipe_id::obj` will show a debug message if the recipe id is not valid (see `string_id<recipe>::obj` in "recipe_dictionary.cpp").